aptos-move/aptos-vm/src/verifier/event_validation.rs

```diff
@@ -139,22 +249,40 @@ pub(crate) fn extract_event_metadata(
     Ok(event_structs)
 }
 
+/// Returns an error if the script uses 0x1::event::emit function (whether as a direct call, or as
+/// a closure). Note that this is not overly restrictive: even a callback to emit an event cannot
+/// be passed, i.e. the following script should fail:
+///
+/// ```move
+/// script {
+///   fun main() {
+///     let f = |e| {
+///       // ... do something here ...
+///       0x1::event::emit(e);
+///     }
+///
+///     // This call creates an event and calls `f` on it.
+///     0x123::some_module::some_function(f);
+///   }
+/// }
+/// ```
+///
+/// This is ok to fail here, as event emission should be done by the module where event is defined.
 pub(crate) fn verify_no_event_emission_in_compiled_script(script: &CompiledScript) -> VMResult<()> {
-    for bc in &script.code().code {
-        if let Bytecode::CallGeneric(index) = bc {
-            let func_instantiation = &script.function_instantiation_at(*index);
-            let func_handle = script.function_handle_at(func_instantiation.handle);
-            let module_handle = script.module_handle_at(func_handle.module);
-            let module_addr = script.address_identifier_at(module_handle.address);
-            let module_name = script.identifier_at(module_handle.name);
-            let func_name = script.identifier_at(func_handle.name);
-            if module_addr == &AccountAddress::ONE
-                && module_name.as_str() == EVENT_MODULE_NAME
-                && func_name.as_str() == EVENT_EMIT_FUNCTION_NAME
-            {
-                return Err(PartialVMError::new(StatusCode::INVALID_OPERATION_IN_SCRIPT)
-                    .finish(Location::Script));
-            }
+    for func_handle in &script.function_handles {
```

Comment:
Nice, looks like instead of scanning the code in the script, you accomplish the same effect (and also cover closures) by looking at loaded external functions.